toonz/sources/image/ffmpeg/tiio_ffmpeg.cpp

```diff
@@ -0,0 +1,338 @@
+	QString tempName = "tempOut%d." + m_intermediateFormat;
+	tempName = m_path.getQString() + tempName;
+
+	QStringList args;
+	args = args + preIArgs;
+	if (!includesInPath) {  //NOTE:  if including the in path, it needs to be in the preIArgs argument.
+		args << "-i";
+		args << tempName;
+	}
+	if (m_hasSoundTrack)
+		args = args + m_audioArgs;
+	args = args + postIArgs;
+	if (overWriteFiles && !includesOutPath) { //if includesOutPath is true, you need to include the overwrite in your postIArgs.
+		args << "-y";
+	}
+	if (!includesOutPath) {
+		args << m_path.getQString();
+	}
+
+	//write the file
+	QProcess ffmpeg;
+	ffmpeg.start(m_ffmpegPath + "/ffmpeg", args);
+	ffmpeg.waitForFinished();
+	QString results = ffmpeg.readAllStandardError();
+	results += ffmpeg.readAllStandardOutput();
+	ffmpeg.close();
+	std::string strResults = results.toStdString();
+}
+
+QString Ffmpeg::runFfprobe(QStringList args) {
+	QProcess ffmpeg;
+	ffmpeg.start(m_ffmpegPath + "/ffprobe", args);
+	ffmpeg.waitForFinished();
+	QString results = ffmpeg.readAllStandardError();
+	results += ffmpeg.readAllStandardOutput();
+	ffmpeg.close();
+	std::string strResults = results.toStdString();
+	return results;
+}
+
+void Ffmpeg::saveSoundTrack(TSoundTrack *st)
+{
+	m_sampleRate = st->getSampleRate();
+	m_channelCount = st->getChannelCount();
+	m_bitsPerSample = st->getBitPerSample();
+	//LONG count = st->getSampleCount();
+	int bufSize = st->getSampleCount() * st->getSampleSize();
+	const UCHAR *buffer = st->getRawData();
+
+	m_audioPath = m_path.getQString() + "tempOut.raw";
+	m_audioFormat = "s" + QString::number(m_bitsPerSample);
+	if (m_bitsPerSample > 8)
+		m_audioFormat = m_audioFormat + "le";
+	std::string strPath = m_audioPath.toStdString();
+
+	QByteArray data;
+	data.insert(0, (char *)buffer, bufSize);
+
+	QFile file(m_audioPath);
+	file.open(QIODevice::WriteOnly);
+	file.write(data);
+	file.close();
+	m_hasSoundTrack = true;
+
+	m_audioArgs << "-f";
+	m_audioArgs << m_audioFormat;
+	m_audioArgs << "-ar";
+	m_audioArgs << QString::number(m_sampleRate);
+	m_audioArgs << "-ac";
+	m_audioArgs << QString::number(m_channelCount);
+	m_audioArgs << "-i";
+	m_audioArgs << m_audioPath;
+
+	//add file to framesWritten for cleanup
+	m_cleanUpList.push_back(m_audioPath);
+}
+
+TRasterImageP Ffmpeg::getImage(int frameIndex) {
+	QString ffmpegCachePath = getFfmpegCache().getQString();
+	QString tempPath = ffmpegCachePath + "//" + QString::fromStdString(m_path.getName()) + QString::fromStdString(m_path.getType());
+	std::string tmpPath = tempPath.toStdString();
+	//QString tempPath= m_path.getQString();
+	QString number = QString("%1").arg(frameIndex, 4, 10, QChar('0'));
+	QString tempName = "In" + number + ".png";
+	tempName = tempPath + tempName;
+
+	//for debugging	
+	std::string strPath = tempName.toStdString();
+	if (TSystem::doesExistFileOrLevel(TFilePath(tempName))) {
+		QImage *temp = new QImage(tempName, "PNG");
+		if (temp){
```

Comment:
In C++, the normal `new` operator will never return `nullptr`, it will throw an exception of type `std::bad_alloc` if it fails. That means there is no need to and no point in testing the return value. 

If the intention was to be sure that the `getImage` function absolutely won't throw no matter what happens, you can use the `std::nothrow` version of the `new` operator instead, but since the function is marked neither `noexcept` nor `throw()`, this is probably not necessary.